# Missing Data
Nosso dataset contém dados sobre reservas feitas em um hotel localizado na cidade de Lisboa, Portugal, entre os anos de 2015 e 2017.

Nosso objetivo é prever se a reserva será cancelada.

Ideia de feature engineering:
-é de portugal?
-é da europa?

In [26]:
#fonte dos dados: https://www.sciencedirect.com/science/article/pii/S2352340918315191
#fonte dos dados: https://github.com/rfordatascience/tidytuesday/tree/master/data/2020/2020-02-11

#importando as bibliotecas:
import pandas as pd
import numpy as np

#importando os dados:
df = pd.read_csv('H2.txt')

#printando um resumo dos dados:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79330 entries, 0 to 79329
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   IsCanceled                   79330 non-null  int64  
 1   LeadTime                     79330 non-null  int64  
 2   ArrivalDateYear              79330 non-null  int64  
 3   ArrivalDateMonth             79330 non-null  object 
 4   ArrivalDateWeekNumber        79330 non-null  int64  
 5   ArrivalDateDayOfMonth        79330 non-null  int64  
 6   StaysInWeekendNights         79330 non-null  int64  
 7   StaysInWeekNights            79330 non-null  int64  
 8   Adults                       79330 non-null  int64  
 9   Children                     79326 non-null  float64
 10  Babies                       79330 non-null  int64  
 11  Meal                         79330 non-null  object 
 12  Country                      79306 non-null  object 
 13  MarketSegment   

### Dicionário de dados

|variable                       |class     |description |
|:------------------------------|:---------|:-----------|
|is_canceled                    |double    | Value indicating if the booking was canceled (1) or not (0) |
|lead_time                      |double    | Number of days that elapsed between the entering date of the booking into the PMS and the arrival date |
|arrival_date_year              |double    | Year of arrival date|
|arrival_date_month             |character | Month of arrival date|
|arrival_date_week_number       |double    | Week number of year for arrival date|
|arrival_date_day_of_month      |double    | Day of arrival date|
|stays_in_weekend_nights        |double    | Number of weekend nights (Saturday or Sunday) the guest stayed or booked to stay at the hotel |
|stays_in_week_nights           |double    |  Number of week nights (Monday to Friday) the guest stayed or booked to stay at the hotel|
|adults                         |double    | Number of adults|
|children                       |double    | Number of children|
|babies                         |double    |Number of babies |
|meal                           |character | Type of meal booked. Categories are presented in standard hospitality meal packages: <br> Undefined/SC – no meal package;<br>BB – Bed & Breakfast; <br> HB – Half board (breakfast and one other meal – usually dinner); <br> FB – Full board (breakfast, lunch and dinner) |
|country                        |character | Country of origin. Categories are represented in the ISO 3155–3:2013 format |
|market_segment                 |character | Market segment designation. In categories, the term “TA” means “Travel Agents” and “TO” means “Tour Operators” |
|distribution_channel           |character | Booking distribution channel. The term “TA” means “Travel Agents” and “TO” means “Tour Operators” |
|is_repeated_guest              |double    | Value indicating if the booking name was from a repeated guest (1) or not (0) |
|previous_cancellations         |double    | Number of previous bookings that were cancelled by the customer prior to the current booking |
|previous_bookings_not_canceled |double    | Number of previous bookings not cancelled by the customer prior to the current booking |
|reserved_room_type             |character | Code of room type reserved. Code is presented instead of designation for anonymity reasons |
|assigned_room_type             |character | Code for the type of room assigned to the booking. Sometimes the assigned room type differs from the reserved room type due to hotel operation reasons (e.g. overbooking) or by customer request. Code is presented instead of designation for anonymity reasons |
|booking_changes                |double    | Number of changes/amendments made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation|
|deposit_type                   |character | Indication on if the customer made a deposit to guarantee the booking. This variable can assume three categories:<br>No Deposit – no deposit was made;<br>Non Refund – a deposit was made in the value of the total stay cost;<br>Refundable – a deposit was made with a value under the total cost of stay. |
|agent                          |character | ID of the travel agency that made the booking |
|company                        |character | ID of the company/entity that made the booking or responsible for paying the booking. ID is presented instead of designation for anonymity reasons |
|days_in_waiting_list           |double    | Number of days the booking was in the waiting list before it was confirmed to the customer |
|customer_type                  |character | Type of booking, assuming one of four categories:<br>Contract - when the booking has an allotment or other type of contract associated to it;<br>Group – when the booking is associated to a group;<br>Transient – when the booking is not part of a group or contract, and is not associated to other transient booking;<br>Transient-party – when the booking is transient, but is associated to at least other transient booking|
|adr                            |double    | Average Daily Rate as defined by dividing the sum of all lodging transactions by the total number of staying nights |
|required_car_parking_spaces    |double    | Number of car parking spaces required by the customer |
|total_of_special_requests      |double    | Number of special requests made by the customer (e.g. twin bed or high floor)|
|reservation_status             |character | Reservation last status, assuming one of three categories:<br>Canceled – booking was canceled by the customer;<br>Check-Out – customer has checked in but already departed;<br>No-Show – customer did not check-in and did inform the hotel of the reason why |
|reservation_status_date        |double    | Date at which the last status was set. This variable can be used in conjunction with the ReservationStatus to understand when was the booking canceled or when did the customer checked-out of the hotel|

fonte: dicionário adaptado do repo que contém o dataset, https://github.com/rfordatascience/tidytuesday/tree/master/data/2020/2020-02-11

Temos 79330 data points e 31 colunas diferentes. Vamos ver quantos data points são nulos, em cada coluna:

In [27]:
#printando a quantidade de dados nulos em cada coluna
df.isna().sum()

IsCanceled                      0
LeadTime                        0
ArrivalDateYear                 0
ArrivalDateMonth                0
ArrivalDateWeekNumber           0
ArrivalDateDayOfMonth           0
StaysInWeekendNights            0
StaysInWeekNights               0
Adults                          0
Children                        4
Babies                          0
Meal                            0
Country                        24
MarketSegment                   0
DistributionChannel             0
IsRepeatedGuest                 0
PreviousCancellations           0
PreviousBookingsNotCanceled     0
ReservedRoomType                0
AssignedRoomType                0
BookingChanges                  0
DepositType                     0
Agent                           0
Company                         0
DaysInWaitingList               0
CustomerType                    0
ADR                             0
RequiredCarParkingSpaces        0
TotalOfSpecialRequests          0
ReservationSta

Temos 4 dados nulos na coluna Children, e 24 nulos em Country. Como representam uma fração muito pequena da quantidade total de dados, vamos eliminar esses data points.

Em seguida, vamos analisar os valores únicos de cada uma das colunas para ver se encontramos missing data.

In [28]:
#eliminando os valores nulos
df.dropna(inplace=True)

#printando os valores únicos de cada coluna, exceto nas colunas ADR e ReservationStatusDate (pois poluiriam visualmente o notebook, pois são muito longas. Vou analisá-las separadamente, em off)
for col in df.drop(['ADR', 'ReservationStatusDate'], axis=1).columns:
    try:
        #algumas colunas possuem valores que têm vários 'espaços' como prefixos. vamos eliminar esses espaços
        df[col] = df[col].apply(lambda x: x.strip(' '))
    except:
        pass
    try:
        print(f"{col}: {sorted(df[col].unique())}")
    except:
        try:
            print(f"{col}: {df[col].unique()}")
        except:
            print(f"ERRO: {col}")
        

IsCanceled: [0, 1]
LeadTime: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 

Pela análise dos valores únicos de cada coluna, mostrado acima, vemos que as colunas Agent e Company possuem valores NULL que não foram lidos como nulos, pois estão em formato de string. Vemos abaixo quantos são:

In [29]:
#quantidade de valores nulos em cada coluna
print(f"valores nulos em Agent: {df.Agent.value_counts().loc['NULL']}")
print(f"valores nulos em Company: {df.Company.value_counts().loc['NULL']}")

valores nulos em Agent: 8121
valores nulos em Company: 75621


Como temos mais de 90% de dados nulos em Company, vamos eliminar essa coluna.

Em Agent, temos cerca de 10% de dados nulos. Manteremos, por enquanto, essa coluna, já que pretendo usar um modelo de Gradient Boosting que é capaz de trabalhar com dados nulos (e.g. xgboost, lightgbm)

Em seguida, vamos verificar se as colunas que possuem valores numéricos estão com datatype correto (e não de string, por exemplo).

In [30]:
#eliminando a coluna Company
df.drop('Company', axis=1, inplace=True)

## Outros missing data
Durante a análise, percebi que existem 324 reservas no dataset que não possuem nenhum dia de hospedagem. Para essas reservas, o custo médio diário foi de 0. Entretanto, 308 delas possuem status de Check-Out.

Não é possível que o hóspede tenha feito a reserva (e o check-out!) sem indicar o dia que ficaria hospedado. Desse modo, suspeito que exista algum erro nesses data points. Por isso, vou eliminá-los da análise.

In [31]:
#existem reservas que não possuem nenhum dia de semana nem de final de semana
#mas o status da maioria delas está como Check-Out
df[(df.StaysInWeekNights == 0) & (df.StaysInWeekendNights == 0)].ReservationStatus.value_counts()

Check-Out    308
Canceled       9
No-Show        7
Name: ReservationStatus, dtype: int64

In [32]:
#As reservas que não possuem nenhum dia de semana nem de final de semana têm custo diário médio de 0
df[(df.StaysInWeekNights == 0) & (df.StaysInWeekendNights == 0)].ADR.value_counts()

0.0    324
Name: ADR, dtype: int64

In [33]:
#eliminando as reservas que não possuem nenhum dia associado

#lista de índices dos data points a serem eliminados
index_list_to_delete = df[(df.StaysInWeekNights == 0) & (df.StaysInWeekendNights == 0)].index

#eliminando esses data points
df.drop(index_list_to_delete, inplace=True)

Outra coisa que percebi é que existem 99 reservas que não possuem adulto, criança ou bebê associado a ela (ou seja, nenhuma pessoa). Como não é possível que isso aconteça, acredito que seja um erro e vou eliminar essas reservas da análise.

In [34]:
#df só com as reservas em que não há pessoas associadas (nenhum adulto, criança ou bebê)
nobody_df = df[(df.Adults == 0) & (df.Children == 0) & (df.Babies == 0)]

print(f'reservas sem qualquer pessoa associada: {len(nobody_df.index)}')

#eliminando esses elementos do df
index_list_to_delete = nobody_df.index
df.drop(index_list_to_delete, inplace=True)

reservas sem qualquer pessoa associada: 99


# Data Types
Abaixo verificamos que nenhuma coluna numérica está com o datatype de string.

In [35]:
#verificando os data types numéricos
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 78879 entries, 0 to 79329
Data columns (total 30 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   IsCanceled                   78879 non-null  int64  
 1   LeadTime                     78879 non-null  int64  
 2   ArrivalDateYear              78879 non-null  int64  
 3   ArrivalDateMonth             78879 non-null  object 
 4   ArrivalDateWeekNumber        78879 non-null  int64  
 5   ArrivalDateDayOfMonth        78879 non-null  int64  
 6   StaysInWeekendNights         78879 non-null  int64  
 7   StaysInWeekNights            78879 non-null  int64  
 8   Adults                       78879 non-null  int64  
 9   Children                     78879 non-null  float64
 10  Babies                       78879 non-null  int64  
 11  Meal                         78879 non-null  object 
 12  Country                      78879 non-null  object 
 13  MarketSegment   

Finalizamos o tratamento inicial dos dados.

Ficamos com um dataset de 78978 pontos e 30 colunas, em que somente a coluna Agent possui valores nulos (cerca de 10%).

Faremos, em seguida, a Análise Exploratória de Dados.

In [36]:
#exportando o dataset tratado
df.to_csv('hotels_tratado.csv', index=False)